lib/spack/spack/compilers/__init__.py

```diff
@@ -255,41 +255,69 @@ def compilers_for_arch(arch_spec, scope=None):
+        self.val = val
+        self.id = id(val)
+
+    def __hash__(self):
+        return self.id
+
+    def __eq__(self, other):
+        return isinstance(other, StrongReference) and self.id == other.id
 
-    if compiler is None:
-        cspec = spack.spec.CompilerSpec(items['spec'])
-        os = items.get('operating_system', None)
-        target = items.get('target', None)
 
-        if not ('paths' in items and
-                all(n in items['paths'] for n in _path_instance_vars)):
-            raise InvalidCompilerConfigurationError(cspec)
+def compiler_from_dict(items):
+    cspec = spack.spec.CompilerSpec(items['spec'])
+    os = items.get('operating_system', None)
+    target = items.get('target', None)
 
-        cls  = class_for_compiler_name(cspec.name)
+    if not ('paths' in items and
+            all(n in items['paths'] for n in _path_instance_vars)):
+        raise InvalidCompilerConfigurationError(cspec)
 
-        compiler_paths = []
-        for c in _path_instance_vars:
-            compiler_path = items['paths'][c]
-            if compiler_path != 'None':
-                compiler_paths.append(compiler_path)
-            else:
-                compiler_paths.append(None)
+    cls  = class_for_compiler_name(cspec.name)
 
-        mods = items.get('modules')
-        if mods == 'None':
-            mods = []
+    compiler_paths = []
+    for c in _path_instance_vars:
+        compiler_path = items['paths'][c]
+        if compiler_path != 'None':
+            compiler_paths.append(compiler_path)
+        else:
+            compiler_paths.append(None)
 
-        alias = items.get('alias', None)
-        compiler_flags = items.get('flags', {})
-        environment = items.get('environment', {})
-        extra_rpaths = items.get('extra_rpaths', [])
+    mods = items.get('modules')
+    if mods == 'None':
+        mods = []
 
-        compiler = cls(cspec, os, target, compiler_paths, mods, alias,
-                       environment, extra_rpaths, **compiler_flags)
-        _compiler_cache[id(items)] = compiler
+    alias = items.get('alias', None)
+    compiler_flags = items.get('flags', {})
+    environment = items.get('environment', {})
+    extra_rpaths = items.get('extra_rpaths', [])
+
+    return cls(cspec, os, target, compiler_paths, mods, alias,
+               environment, extra_rpaths, **compiler_flags)
+
+
+def _compiler_from_config_entry(items):
+    """Note this is intended for internal use only. To avoid re-parsing
+       the same config dictionary this keeps track of its location in
+       memory. If you provide the same dictionary twice it will return
+       the same Compiler object (regardless of whether the dictionary
+       entries have changed).
+    """
+    config_id = StrongReference(items)
+    compiler = _compiler_cache.get(config_id, None)
+
+    if compiler is None:
+        compiler = compiler_from_dict(items)
+        _compiler_cache[StrongReference(items)] = compiler
```

Comment:
I can see the value of using `StrongReference` instances for the key -- you're still taking advantage of the unique value from `id()` and you hang onto a reference to `items` -- but I'm not following why you need to instantiate another object at line 320 (versus using the one at line 315).

Reply:
> I'm not following why you need to instantiate another object at line 320 (versus using the one at line 315).

It is not required (although also not wrong). I'm not sure why I did it that way before: I can change it.

Reply:
I know it isn't wrong but it is also unnecessary and, IMO, a potential source of confusion for people reading the comments.